apps/cli/src/commands/boot-challenger.ts

```diff
@@ -134,11 +134,13 @@ const startListener = async (commandInstance: Vorpal.CommandInstance) => {
                     if (errorCount > NUM_CON_WS_ALLOWED_ERRORS) {
                         stopListener(listener);
                         resolve(error);
+                    } else {
+                        //If the websocket just reconnected automatically we only want to try to re-post the last possibly missed challenge
+                        await processMissedAssertions(commandInstance);
```

Comment:
I think there’s a few things about this solution that won’t work.
- When we call processMissedAssertions, this is skipped on the last WS error, this shouldn’t exist inside of the if else, and should just exist before the `if (errorCount > NUM_CON_WS_ALLOWED_ERRORS)`
- Calling processMissedAssertions as a result of the listenForAssertions callback is going to create race conditions
    - If the processMissedAssertions is running, and then another error comes in, then it will run processMissedAssertions more than once
    - If we call processMissedAssertions, and then a final error comes in, that final error will close the runtime while we’re running processMissedAssertions

Reply:
Added a catch in the callback to not stop the runntime if that should fail.

About calling in the else, this should catch the websocket reconnecting at the exact time of a challenge.

Reply:
```mermaid
sequenceDiagram
    participant E as Error Listener
    participant A as Instance A of processMissedAssertions
    participant B as Instance B of processMissedAssertions
    participant C as Instance C of processMissedAssertions

    Note over E: Error Occurs
    E->>+A: Trigger Instance A
    Note over A: Processing...

    Note over E: Another Error Occurs
    E->>+B: Trigger Instance B
    Note over B: Processing...

    Note over E: Yet Another Error Occurs
    E->>+C: Trigger Instance C
    Note over C: Processing...

    Note over A: Processing Done
    A-->>-E: Complete

    Note over B: Processing Done
    B-->>-E: Complete

    Note over C: Processing Done
    C-->>-E: Complete
```

What you are saying is correct, but the issue is this results in up to `NUM_CON_WS_ALLOWED_ERRORS` instances of `processMissedAssertions` running in parralel, which is very heavy. 

I would suggest the following...

We should call processMissedAssertions when `NUM_CON_WS_ALLOWED_ERRORS`  is exceeded, so it shuts down the listener and only runs once. Then if we recover before `NUM_CON_WS_ALLOWED_ERRORS`  is exceeded, we check to see if `errorCount > 1` and then run `processMissedAssertions` in that case. In this case, per connection error `processMissedAssertions` is only called once.

```mermaid
flowchart TB
    A[Start Monitoring for Errors] --> B{Error Detected?}
    B -->|Yes| C{Error Count > NUM_CON_WS_ALLOWED_ERRORS?}
    B -->|No| A
    C -->|Yes| D[Shut Down Listener]
    C -->|No| E[Increment Error Count]
    D --> F[Call processMissedAssertions]
    E --> G{Recover Before Error Count Exceeds Limit?}
    F --> H[End]
    G -->|Yes| I{Error Count > 1?}
    G -->|No| A
    I -->|Yes| J[Call processMissedAssertions]
    I -->|No| A
    J --> A
```

Reply:
That makes sense, added the check for errorCount != 0 so when the WS auto reconnects successful it will only once check for the missed assertion.
If it fails `NUM_CON_WS_ALLOWED_ERRORS` in a row, the listener will restart and check for missed assertions.